op_vault.go

```diff
@@ -0,0 +1,173 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"strings"
+)
+
+// The VaultOperator provides a means of injecting credentials and
+// other secrets from a Vault (vaultproject.io) Secure Key Storage
+// instance.
+type VaultOperator struct{}
+
+// Setup ...
+func (VaultOperator) Setup() error {
+	return nil
+}
+
+// Phase identifies what phase of document management the vault
+// operator should be evaulated in.  Vault lives in the Eval phase
+func (VaultOperator) Phase() OperatorPhase {
+	return EvalPhase
+}
+
+// Dependencies collects implicit dependencies that a given
+// `(( vault ... ))` call has.  There are none.
+func (VaultOperator) Dependencies(_ *Evaluator, _ []*Expr, _ []*Cursor) []*Cursor {
+	return []*Cursor{}
+}
+
+// Run executes the `(( vault ... ))` operator call, which entails
+// interacting with the (unsealed) Vault instance to retrieve the
+// given secrets.
+func (VaultOperator) Run(ev *Evaluator, args []*Expr) (*Response, error) {
+	DEBUG("running (( vault ... )) operation at $.%s", ev.Here)
+	defer DEBUG("done with (( vault ... )) operation at $.%s\n", ev.Here)
+
+	// syntax: (( vault "secret/path:key" ))
+	if len(args) != 1 {
+		return nil, fmt.Errorf("vault operator requires at least one argument")
+	}
+
+	v, err := args[0].Resolve(ev.Tree)
+	if err != nil {
+		DEBUG("  arg[0]: failed to resolve expression to a concrete value")
+		DEBUG("     [0]: error was: %s", err)
+		return nil, err
+	}
+
+	var key string
+	switch v.Type {
+	case Literal:
+		DEBUG("  arg[0]: using string literal '%v'", v.Literal)
+		key = fmt.Sprintf("%v", v.Literal)
+
+	case Reference:
+		DEBUG("  arg[0]: trying to resolve reference $.%s", v.Reference)
+		s, err := v.Reference.Resolve(ev.Tree)
+		if err != nil {
```

Comment:
I thought the original Resolve() resolved things so that v would always be a literal string/map/array, so that operators don't need to check for/perform recursive resolving. Is that not the case?